source/common/network/listener_filter_buffer_impl.h

```diff
@@ -0,0 +1,84 @@
+#pragma once
+
+#include <atomic>
+#include <functional>
+#include <memory>
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/network/io_handle.h"
+#include "envoy/network/listener_filter_buffer.h"
+
+#include "source/common/buffer/buffer_impl.h"
+
+namespace Envoy {
+namespace Network {
+
+// TODO(soulxu): check whether provide the error code to the callback.
+using ListenerFilterBufferOnCloseCb = std::function<void()>;
+using ListenerFilterBufferOnDataCb = std::function<void()>;
+
+enum class PeekState {
+  // Peek data status successful.
+  Done,
+  // Need to try again.
+  Again,
+  // Error to peek data.
+  Error
+};
+
+class ListenerFilterBufferImpl : public ListenerFilterBuffer, Logger::Loggable<Logger::Id::filter> {
+public:
+  ListenerFilterBufferImpl(IoHandle& io_handle, Event::Dispatcher& dispatcher,
+                           ListenerFilterBufferOnCloseCb close_cb,
+                           ListenerFilterBufferOnDataCb on_data_cb, uint64_t buffer_size)
+      : io_handle_(io_handle), dispatcher_(dispatcher), on_close_cb_(close_cb),
+        on_data_cb_(on_data_cb), buffer_(new Buffer::OwnedImpl()) {
+    // If the buffer_size not greater than 0, it means that doesn't expect any data.
+    ASSERT(buffer_size > 0);
+    // Since we need the single slice for peek from the socket, so initialize that
+    // single slice.
+    auto reservation = buffer_->reserveSingleSlice(buffer_size);
+    reservation.commit(buffer_size);
+    io_handle_.initializeFileEvent(
+        dispatcher_, [this](uint32_t events) { onFileEvent(events); },
+        Event::PlatformDefaultTriggerType,
+        Event::FileReadyType::Read | Event::FileReadyType::Closed);
+  }
+
+  const Buffer::ConstRawSlice rawSlice() const override;
```

Comment:
Add a comment `// ListenerFilterBuffer` to indicate these came from the interface.

Reply:
got it